### DISCUSSION TITLE
Records won't warn about an unimplemented abstract ToString

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#8907] [General]
**Version Used**:  https://github.com/dotnet/roslyn/commit/98ea496177bc8e607dbf454abd6b5a5e4678aed2

**Steps to Reproduce**:

```cs
abstract class ClassBase
{
    public abstract override string ToString();
    sealed class Class : ClassBase; // error
}
abstract record RecordBase
{
    public abstract override string ToString();
    sealed record Record : RecordBase; // ok (error expected)
}

```

**Expected Behavior**: `error CS0534: 'RecordBase.Record' does not implement inherited abstract member 'RecordBase.ToString()'`

**Actual Behavior**: there's no error on the record declaration.

**Comment by CyrusNajmabadi** [comment]
Why should this be an error?  The compiler synthesizes ToString for all records, right?  So the method will be implemented.

**Comment by alrz** [comment]
I'd interpret abstract to mean "must be user provided in source", so a compiler-generated method shouldn't satisfy that.

**Comment by CyrusNajmabadi** [comment]
Abstract means an override must be provided. The compiler provides that override. It has nothing to do with the user providing it or not :-)

**Comment by alrz** [comment]
> Abstract means an override must be provided. The compiler provides that override. It has nothing to do with the user providing it or not :-)

And that's specified by the CIL. There's a lot of things that metadata doesn't care about but the language does. Like out/ref for example. I think this could be adjacent to that, only if you agree that it should be possible to reabstract record members that are inherited from object.

**Comment by CyrusNajmabadi** [comment]
I don't know what you're asking for here.  You said you thought this was a bug.  I'm saying that i don't think this is a bug in the compiler, and the compiler is doing exactly the right thing as far as the language is concerned.  At no point in the design of records did we say that if the base method is abstract that you would then have an error if the user didn't implement the method in source themselves.

If you want that, you'd need to propose a language change over at csharplang.

**Reply by alrz** [reply]
Thanks. I could not transfer to csharplang though.

**Comment by jnm2** [comment]
Obligatory "this could be an analyzer." Also, as a type author you can seal ToString in the base class (https://github.com/dotnet/csharplang/issues/4174) and have it call a new protected abstract member.
